File: 2405-optimal-partition-of-string/2405-optimal-partition-of-string.java
class Solution {
    public int partitionString(String s) {
        
        Set<Character> set = new HashSet();
        int count = 0;
        
        for(int j=0; j<s.length(); j++){
            if(set.contains(s.charAt(j))){
                count++;
                set.clear();
            }
            set.add(s.charAt(j));
        }
        
        return (set.size() > 0)?(count+1):count;
        
    }
}

/*

a - 6
b - 5
c - 3



0123456
abacaba
i
j

*/

File: 2405-optimal-partition-of-string/2405-optimal-partition-of-string.py
class Solution:
    def partitionString(self, s: str) -> int:
        
        def default_zero():
            return 0
        j = 0
        freq = defaultdict(default_zero)
        count = 0
        
        for i in range(0,len(s)):
            if(freq[s[i]] == 1):
                freq[s[i]]+=1
                count+=1
                while j < i:
                    freq[s[j]]-=1
                    j+=1
            else:
                freq[s[i]]+=1
        
        return count+1
                    
            
        
        